File: Stack/Min Stack/MinStack.go
type StackElement struct {
	Value     int
	MinValue  int
}

type MinStack struct {
	minStack *arraystack.Stack
}

func Constructor() MinStack {
	return MinStack{
		minStack: arraystack.New(),
	}
}

func (ms *MinStack) Push(val int) {
	if ms.minStack.Empty() {
		ms.minStack.Push(StackElement{Value: val, MinValue: val})
	} else {
		topElem, _ := ms.minStack.Peek()
		newMin := int(math.Min(float64(val), float64(topElem.(StackElement).MinValue)))
		ms.minStack.Push(StackElement{Value: val, MinValue: newMin})
	}
}

func (ms *MinStack) Pop() {
	if !ms.minStack.Empty() {
		ms.minStack.Pop()
	}
}

func (ms *MinStack) Top() int {
	if topElem, ok := ms.minStack.Peek(); ok {
		return topElem.(StackElement).Value
	}
	return 0
}

func (ms *MinStack) GetMin() int {
	if topElem, ok := ms.minStack.Peek(); ok {
		return topElem.(StackElement).MinValue
	}
	return 0
}

File: Stack/Min Stack/MinStack.cs
public class MinStack
{
    private class StackElement
    {
        public int Value { get; }
        public int MinValue { get; }

        public StackElement(int value, int minValue)
        {
            Value = value;
            MinValue = minValue;
        }
    }

    private Stack<StackElement> minStack;

    public MinStack()
    {
        minStack = new Stack<StackElement>();
    }

    public void Push(int val)
    {
        if (minStack.Count == 0)
        {
            minStack.Push(new StackElement(val, val));
        }
        else
        {
            int newMin = Math.Min(val, minStack.Peek().MinValue);
            minStack.Push(new StackElement(val, newMin));
        }
    }

    public void Pop()
    {
        if (minStack.Count > 0)
        {
            minStack.Pop();
        }
    }

    public int Top()
    {
        return minStack.Peek().Value;
    }

    public int GetMin()
    {
        return minStack.Peek().MinValue;
    }
}